src/kokkos/plugin-PixelTriplets/kokkos/CAHitNtupletGeneratorKernels.cc

```diff
@@ -191,37 +191,51 @@ namespace KOKKOS_NAMESPACE {
     if (0 == nhits)
       return;  // protect against empty events
 
-      // FIXME avoid magic numbers
-#ifdef TODO
+    // FIXME avoid magic numbers
     auto nActualPairs = gpuPixelDoublets::nPairs;
     if (!m_params.includeJumpingForwardDoublets_)
       nActualPairs = 15;
     if (m_params.minHitsPerNtuplet_ > 3) {
       nActualPairs = 13;
     }
-#endif
 
-#ifdef TODO
     assert(nActualPairs <= gpuPixelDoublets::nPairs);
+#ifdef KOKKOS_BACKEND_CUDA
     int stride = 4;
-    int threadsPerBlock = gpuPixelDoublets::getDoubletsFromHistoMaxBlockSize / stride;
-    int blocks = (4 * nhits + threadsPerBlock - 1) / threadsPerBlock;
-    dim3 blks(1, blocks, 1);
-    dim3 thrs(stride, threadsPerBlock, 1);
-    gpuPixelDoublets::getDoubletsFromHisto<<<blks, thrs, 0, stream>>>(device_theCells_.get(),
-                                                                      device_nCells_,
-                                                                      device_theCellNeighbors_,
-                                                                      device_theCellTracks_,
-                                                                      hh.view(),
-                                                                      device_isOuterHitOfCell_.get(),
-                                                                      nActualPairs,
-                                                                      m_params.idealConditions_,
-                                                                      m_params.doClusterCut_,
-                                                                      m_params.doZ0Cut_,
-                                                                      m_params.doPtCut_,
-                                                                      m_params.maxNumberOfDoublets_);
-    cudaCheck(cudaGetLastError());
+    teamSize = gpuPixelDoublets::getDoubletsFromHistoMaxBlockSize / stride;
+    leagueSize = (4 * nhits + teamSize - 1) / teamSize;
+    Kokkos::TeamPolicy<KokkosExecSpace,
+                       Kokkos::LaunchBounds<gpuPixelDoublets::getDoubletsFromHistoMaxBlockSize,
+                                            gpuPixelDoublets::getDoubletsFromHistoMinBlocksPerMP>>
+        tempPolicy{execSpace, leagueSize, teamSize * stride};
+#else
+    int stride = 1;
+    Kokkos::TeamPolicy<KokkosExecSpace,
+                       Kokkos::LaunchBounds<gpuPixelDoublets::getDoubletsFromHistoMaxBlockSize,
+                                            gpuPixelDoublets::getDoubletsFromHistoMinBlocksPerMP>>
+        tempPolicy{execSpace, 1, Kokkos::AUTO()};
```

Comment:
Same here.

Reply:
Here to have the same flexibility is less straightforward. Cause current multi-team kernels are basically a copy-paste from CUDA implementations which sometimes use CUDA-specific algorithms. In this `getDoubletsFromHisto`, for example, it set the blockDim as 64 (or 128 I don't remember) and hardcoded `stride size = 4`. In a multi-threading use case on CPU, it's possible that we have less than 4 hardware threads or we launch an odd number of threads. Also, I need to investigate if the code is flexible enough to use a different stride size other than 4. Just FYI, all my previous porting had the same issue also since I thought we only need to take care of CUDA and Serial.

Reply:
Thanks for the explanation. Let's keep it like this then, but could you add the explanation in a comment in the file itself (once in the first occurrence is enough)?

Reply:
done